queries.yaml

```diff
@@ -975,19 +958,77 @@ post_process:
   - fn: vectordatasource.transform.drop_properties_with_prefix
     params: {prefix: mz_}
 
+  # drop small inners in buildings before we do the merge.
+  - fn: vectordatasource.transform.drop_small_inners
+    params:
+      end_zoom: 15
+      source_layers: [buildings]
+      pixel_area: 0.25
+
   - fn: vectordatasource.transform.merge_building_features
     params:
       source_layer: buildings
       start_zoom: 13
-      end_zoom: 15
+      end_zoom: 16
       quantize:
-        13: vectordatasource.transform.quantize_height_round_nearest_10_meters
-        14: vectordatasource.transform.quantize_height_round_nearest_5_meters
+        13: vectordatasource.transform.quantize_height_round_nearest_20_meters
+        14: vectordatasource.transform.quantize_height_round_nearest_10_meters
+        15: vectordatasource.transform.quantize_height_round_nearest_10_meters
+      # todo: keep instead of drop?
       drop:
         - name
+        - short_name
         - addr_housenumber
         - addr_street
-      max_merged_features: 1000
+        - osm_relation
+      # NOTE: max_merged_features is set to keep the time taken for geometry
+      # merging down, as it seems to go up with the square of the number of
+      # features.
+      max_merged_features: 2000
+
+  # drop small inners in buildings after we do the merge, too! because the merge
+  # process may also have created small inners.
+  - fn: vectordatasource.transform.drop_small_inners
+    params:
+      end_zoom: 15
+      source_layers: [buildings]
+      pixel_area: 0.25
+
+  # assign `scale_rank` - AFTER merging, as this changes the area and volume
+  - fn: vectordatasource.transform.csv_match_properties
+    resources:
+      matcher:
+        type: file
+        init_fn: vectordatasource.transform.CSVMatcher
+        path: spreadsheets/scale_rank/buildings.csv
+    params:
+      source_layer: buildings
+      target_value_type: int
+
+  - fn: vectordatasource.transform.numeric_min_filter
+    params:
+      source_layer: buildings
+      mode: any
+      filters:
+        13: { area: 5000, height: 150, volume: 150000 }
+        14: { area: 3000, height: 100, volume: 100000 }
+        15: { area:  500, height: 100, volume: 100000 }
+
+  # drop features not matching scale rank requirements (is is a max, so couldn't
+  # be merged with the preceding numeric min).
+  - fn: vectordatasource.transform.drop_features_where
+    params:
+      source_layer: buildings
+      start_zoom: 13
+      end_zoom: 14
+      where: scale_rank > 2
+
+  - fn: vectordatasource.transform.drop_features_where
+    params:
+      source_layer: buildings
+      start_zoom: 14
+      end_zoom: 15
+      where: scale_rank > 3
```

Comment:
This makes sense to drop afterwards, but is it crazy to initially drop some of the smaller buildings too to cut down on the candidates we need to consider for merging? Or is this actually counterproductive because it might cut down on the merging that we can actually do? Does the drop smaller inners handle this well enough anyway?

Reply:
It's more "coherent" to operate on the entire urban fabric... but agree it can result in large aggregate polygons that would otherwise have been dropped as smaller inputs. Generally they'll still have low height and can be filtered client-side, or we could later drop them server-side if we need more file size savings.